workflow/src/main/java/bio/terra/tanagra/workflow/BuildPathsForHierarchy.java

```diff
@@ -0,0 +1,325 @@
+                      Iterator<String> pathTagIter = element.getValue().getAll(pathTag).iterator();
+                      Iterator<String> parentTagIter =
+                          element.getValue().getAll(parentTag).iterator();
+
+                      // there may be multiple possible next steps (i.e. the current node may have
+                      // multiple parents). just pick the first one
+                      String nextNodeInPath = parentTagIter.hasNext() ? parentTagIter.next() : null;
+
+                      // iterate through all the paths that need this relationship to complete their
+                      // next step
+                      while (pathTagIter.hasNext()) {
+                        String currentPath = pathTagIter.next();
+
+                        if (nextNodeInPath == null) {
+                          // if there are no relationships to complete the next step, then we've
+                          // reached a root node. just keep the path as is and move on the next one
+                          context.output(KV.of(node, currentPath));
+                        } else {
+                          // if there is a next node, then append it to the path and make it the new
+                          // key
+                          context.output(
+                              KV.of(
+                                  Long.valueOf(nextNodeInPath),
+                                  currentPath + "." + nextNodeInPath));
+                        }
+                      }
+                    }
+                  }));
+    }
+    // example for CoGroupByKey+ParDo iteration above:
+    //    desired result (node,path)
+    //    (a1,"a1.a2.a3")
+    //    (a2,"a2.a3")
+    //    (a3,"a3")
+    //    (b1,"b1.b2")
+    //    (b2,"b2")
+    //    (c1,"c1")
+    //
+    //    childParentKVs (child,parent)
+    //    (a1,"a2")
+    //    (a2,"a3")
+    //    (b1,"b2")
+    //
+    //    [iteration 0] nodePathKVs (next node,path)
+    //    (a1,"a1")
+    //    (a2,"a2")
+    //    (a3,"a3")
+    //    (b1,"b1")
+    //    (b2,"b2")
+    //    (c1,"c1")
+    //    [iteration 0] nodeParentPathJoin (next node,paths,next node parents)
+    //    (a1,["a1"],["a2"])
+    //    (a2,["a2"],["a3"])
+    //    (a3,["a3"],[])
+    //    (b1,["b1"],["b2"])
+    //    (b2,["b2"],[])
+    //    (c1,["c1"],[])
+    //
+    //    [iteration 1] nodePathKVs (next node,path)
+    //    (a2,"a2.a1")
+    //    (a3,"a3.a2")
+    //    (a3,"a3")
+    //    (b2,"b2.b1")
+    //    (b2,"b2")
+    //    (c1,"c1")
+    //    [iteration 1] nodeParentPathJoin (next node,paths,next node parents)
+    //    (a1,[],["a2"])
+    //    (a2,["a2.a1"],["a3"])
+    //    (a3,["a3.a2","a3"],[])
+    //    (b1,[],["b2]")
+    //    (b2,["b2.b1","b2"],[])
+    //    (c1,["c1"],[])
+
+    // swap the key of all pairs from the next node in the path (which should all be root nodes at
+    // this point), to the first node in the path
```

Comment:
nit: and trim the target node from the path?

Reply:
Right, added.